### ISSUE TITLE
Cursor offset when scaling one screen in wayland (KDE Plasma)

### ISSUE THREAD

**Issue by RedstoneLP2** [galister/X11Overlay#13]
I have two monitors connected to my PC, one of which is scaled to 85% using the scale setting of the Display configuration in KDE Plasma. 
When Enabling this screen in WlxOverlay the cursor still seems to use the coordinates as if the screen was scaled to 100%, making it impossible to move the cursor to the bottom or right side of the screen.

My best guess (without looking at the code) would be that WlxOverlay uses the resolution to map the cursor position and Plasma expects it to include scaling

**Comment by galister** [comment]
WlxOverlay uses the logical size of each screen as reported by `xdg_output_v1`

https://github.com/galister/WlxOverlay/blob/b101607ea122fd265d1d11d431a691d429ed40ec/Desktop/Wayland/WaylandOutput.cs#L32-L35

https://github.com/galister/WlxOverlay/blob/b101607ea122fd265d1d11d431a691d429ed40ec/Desktop/Wayland/WaylandOutput.cs#L48-L70

It does work for me with scaled screens on wlroots, so not sure that's different here. mind posting a `wayland-info`?

**Comment by RedstoneLP2** [comment]
```
╭─redstonelp2@Manjaro-Ryzen3-1200 ~ 
╰─$ wayland-info
interface: 'wl_compositor',                              version:  5, name:  1
interface: 'zwp_tablet_manager_v2',                      version:  1, name:  3
interface: 'zwp_keyboard_shortcuts_inhibit_manager_v1',  version:  1, name:  4
interface: 'xdg_wm_base',                                version:  4, name:  6
interface: 'zwlr_layer_shell_v1',                        version:  3, name:  7
interface: 'zxdg_decoration_manager_v1',                 version:  1, name:  8
interface: 'wp_viewporter',                              version:  1, name:  9
interface: 'wp_fractional_scale_manager_v1',             version:  1, name: 10
interface: 'wl_shm',                                     version:  1, name: 11
	formats (fourcc):
	0x38344258 = 'XB48'
	0x38344241 = 'AB48'
	0x30334258 = 'XB30'
	0x30334241 = 'AB30'
	0x30335258 = 'XR30'
	0x30335241 = 'AR30'
	         1 = 'XR24'
	         0 = 'AR24'
interface: 'wl_seat',                                    version:  8, name: 12
	name: 
	capabilities: pointer keyboard touch
	keyboard repeat rate: 25
	keyboard repeat delay: 660
interface: 'zwp_pointer_gestures_v1',                    version:  3, name: 13
interface: 'zwp_pointer_constraints_v1',                 version:  1, name: 14
interface: 'zwp_relative_pointer_manager_v1',            version:  1, name: 15
interface: 'wl_data_device_manager',                     version:  3, name: 16
interface: 'zwlr_data_control_manager_v1',               version:  2, name: 17
interface: 'zwp_primary_selection_device_manager_v1',    version:  1, name: 18
interface: 'org_kde_kwin_idle',                          version:  1, name: 19
interface: 'zwp_idle_inhibit_manager_v1',                version:  1, name: 20
interface: 'ext_idle_notifier_v1',                       version:  1, name: 21
interface: 'org_kde_plasma_shell',                       version:  8, name: 22
interface: 'org_kde_kwin_appmenu_manager',               version:  1, name: 23
interface: 'org_kde_kwin_server_decoration_palette_manager', version:  1, name: 24
interface: 'org_kde_plasma_virtual_desktop_management',  version:  2, name: 26
interface: 'org_kde_kwin_shadow_manager',                version:  2, name: 28
interface: 'org_kde_kwin_dpms_manager',                  version:  1, name: 29
interface: 'org_kde_kwin_server_decoration_manager',     version:  1, name: 30
interface: 'kde_output_management_v2',                   version:  3, name: 31
interface: 'zxdg_output_manager_v1',                     version:  3, name: 32
	xdg_output_v1
		output: 49
		name: 'HDMI-A-1'
		description: 'Samsung Electric Company HDMI-A-1-SAMSUNG'
		logical_x: 1882, logical_y: 0
		logical_width: 1920, logical_height: 1080
	xdg_output_v1
		output: 48
		name: 'DP-2'
		description: 'Acer Technologies G206HQL/LZ4EE0012413'
		logical_x: 0, logical_y: 11
		logical_width: 1882, logical_height: 1059
interface: 'wl_subcompositor',                           version:  1, name: 33
interface: 'zxdg_exporter_v2',                           version:  1, name: 34
interface: 'zxdg_importer_v2',                           version:  1, name: 35
interface: 'xdg_activation_v1',                          version:  1, name: 38
interface: 'wp_content_type_manager_v1',                 version:  1, name: 40
interface: 'wp_tearing_control_manager_v1',              version:  1, name: 41
interface: 'wl_drm',                                     version:  2, name: 43
interface: 'zwp_linux_dmabuf_v1',                        version:  4, name: 44
	main device: 0xE200
	tranche
		target device: 0xE200
		flags: none
		formats (fourcc) and modifiers (names):
		0x20363152 = 'R16 '; 0x00ffffffffffffff = INVALID
		0x48344241 = 'AB4H'; 0x00ffffffffffffff = INVALID
		0x38344241 = 'AB48'; 0x00ffffffffffffff = INVALID
		0x56555958 = 'XYUV'; 0x00ffffffffffffff = INVALID
		0x56595559 = 'YUYV'; 0x00ffffffffffffff = INVALID
		0x39555659 = 'YVU9'; 0x00ffffffffffffff = INVALID
		0x3631564e = 'NV16'; 0x00ffffffffffffff = INVALID
		0x36313259 = 'Y216'; 0x00ffffffffffffff = INVALID
		0x32335247 = 'GR32'; 0x00ffffffffffffff = INVALID
		0x3231564e = 'NV12'; 0x00ffffffffffffff = INVALID
		0x32313259 = 'Y212'; 0x00ffffffffffffff = INVALID
		0x30313050 = 'P010'; 0x00ffffffffffffff = INVALID
		0x36313459 = 'Y416'; 0x00ffffffffffffff = INVALID
		0x39565559 = 'YUV9'; 0x00ffffffffffffff = INVALID
		0x35315241 = 'AR15'; 0x00ffffffffffffff = INVALID
		0x34325559 = 'YU24'; 0x00ffffffffffffff = INVALID
		0x56555941 = 'AYUV'; 0x00ffffffffffffff = INVALID
		0x34325659 = 'YV24'; 0x00ffffffffffffff = INVALID
		0x32313459 = 'Y412'; 0x00ffffffffffffff = INVALID
		0x31315559 = 'YU11'; 0x00ffffffffffffff = INVALID
		0x31315659 = 'YV11'; 0x00ffffffffffffff = INVALID
		0x36315559 = 'YU16'; 0x00ffffffffffffff = INVALID
		0x30313259 = 'Y210'; 0x00ffffffffffffff = INVALID
		0x36315659 = 'YV16'; 0x00ffffffffffffff = INVALID
		0x32315559 = 'YU12'; 0x00ffffffffffffff = INVALID
		0x36313050 = 'P016'; 0x00ffffffffffffff = INVALID
		0x32315659 = 'YV12'; 0x00ffffffffffffff = INVALID
		0x36314752 = 'RG16'; 0x00ffffffffffffff = INVALID
		0x30333050 = 'P030'; 0x00ffffffffffffff = INVALID
		0x20203852 = 'R8  '; 0x00ffffffffffffff = INVALID
		0x48344258 = 'XB4H'; 0x00ffffffffffffff = INVALID
		0x38344258 = 'XB48'; 0x00ffffffffffffff = INVALID
		0x38385247 = 'GR88'; 0x00ffffffffffffff = INVALID
		0x32313050 = 'P012'; 0x00ffffffffffffff = INVALID
		0x30313459 = 'Y410'; 0x00ffffffffffffff = INVALID
		0x59565955 = 'UYVY'; 0x00ffffffffffffff = INVALID
	tranche
		target device: 0xE200
		flags: none
		formats (fourcc) and modifiers (names):
		0x34324241 = 'AB24'; 0x00ffffffffffffff = INVALID
		0x34325258 = 'XR24'; 0x00ffffffffffffff = INVALID
		0x34324258 = 'XB24'; 0x00ffffffffffffff = INVALID
		0x34325241 = 'AR24'; 0x00ffffffffffffff = INVALID
	tranche
		target device: 0xE200
		flags: none
		formats (fourcc) and modifiers (names):
		0x30335241 = 'AR30'; 0x00ffffffffffffff = INVALID
		0x30334241 = 'AB30'; 0x00ffffffffffffff = INVALID
		0x30335258 = 'XR30'; 0x00ffffffffffffff = INVALID
		0x30334258 = 'XB30'; 0x00ffffffffffffff = INVALID
interface: 'kde_output_device_v2',                       version:  2, name: 46
interface: 'kde_output_device_v2',                       version:  2, name: 47
interface: 'wl_output',                                  version:  4, name: 48
	name: DP-2
	description: Acer Technologies G206HQL/LZ4EE0012413
	x: 0, y: 11, scale: 1,
	physical_width: 430 mm, physical_height: 240 mm,
	make: 'Acer Technologies', model: 'G206HQL/LZ4EE0012413',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 1600 px, height: 900 px, refresh: 60.000 Hz,
		flags: current
interface: 'wl_output',                                  version:  4, name: 49
	name: HDMI-A-1
	description: Samsung Electric Company HDMI-A-1-SAMSUNG
	x: 1882, y: 0, scale: 1,
	physical_width: 480 mm, physical_height: 270 mm,
	make: 'Samsung Electric Company', model: 'HDMI-A-1-SAMSUNG',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 1920 px, height: 1080 px, refresh: 60.000 Hz,
		flags: current
interface: 'wp_drm_lease_device_v1',                     version:  1, name: 50
interface: 'kde_output_order_v1',                        version:  1, name: 51
interface: 'zwp_text_input_manager_v1',                  version:  1, name: 52
interface: 'zwp_text_input_manager_v2',                  version:  1, name: 53
interface: 'zwp_text_input_manager_v3',                  version:  1, name: 54
interface: 'org_kde_kwin_contrast_manager',              version:  2, name: 56
interface: 'org_kde_kwin_blur_manager',                  version:  1, name: 57
interface: 'org_kde_kwin_slide_manager',                 version:  1, name: 58
```
The display in question should be DP-2.
Looks like Plasma does scaling differently from wlroots (if the "scale: 1" is any indication)
If you need anything else, I'm glad to help

**Comment by galister** [comment]
Out of curiosity, does the cursor match perfectly if you scale the display up (>1) instead of down?

**Comment by RedstoneLP2** [comment]
nope, using the KDE Settings to change it to 120% the cursor is offset into the other direction.
also: i noticed it's possible to select the "wrong" screen (might open a different issue for that)

**Comment by galister** [comment]
There's no way to correlate a pipewire capture to a WlOutput, so it's your responsibility to select the correct screen. even though the pipewire protocol supports sending both screen position and resolution, Gnome is known to not send position and KDE sends the position zeroed out, regardless of the actual logical position of the screen. As such the only way we know on the pipewire side is the resolution, which is not a unique identifier.

I'm not sure, maybe for KDE we need to use the physical resolution for uinput, not the logical one. 
Can you try changing this to `var size = Size;`, and see if that helps?

https://github.com/galister/WlxOverlay/blob/b101607ea122fd265d1d11d431a691d429ed40ec/Desktop/Wayland/WaylandOutput.cs#L50

**Comment by RedstoneLP2** [comment]
Somewhat unrelated: I never managed to get it to run when building myself

```
╭─redstonelp2@Manjaro-Ryzen3-1200 ~/GIT/WlxOverlay/bin/Debug/net6.0 ‹master› 
╰─$ ./WlxOverlay 
WlxOverlay v1.3.1-3-gb101607
OpenVR Version: 1.25.8
IVROverlay_025: pass
IVRCompositor_027: pass
IVRInput_010: pass
Loaded 9 input actions for /actions/default
export_inputs will not be used.
HMD running @ 90 Hz
Wayland detected.
EGL 1.5 initialized.
Using PipeWire capture.
PipeWire version: 0.3.71
Using DMA-Buf with supported formats: DRM_FORMAT_ARGB8888, DRM_FORMAT_ABGR8888, DRM_FORMAT_XRGB8888, DRM_FORMAT_XBGR8888
 You will be prompted one screen at a time.
 Please select the corresponding screen on the prompt.
 Cancel the prompt if you do not wish to capture the given screen. 
 If your compositor supports org.freedesktop.portal.ScreenCast v4, you will only be prompted once.
HDMI-A-1 -> 193
DP-2 -> 192
Using paplay for audio output.
Listening for notifications @ 127.0.0.1:42069
GL Context initialized
Unhandled exception. System.Collections.Generic.KeyNotFoundException: The given key 'SpaceDrag' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at WlxOverlay.Overlays.LaserPointer.EvaluateInput() in /mnt/1TB-SSD/home/redstonelp2/GIT/WlxOverlay/Overlays/LaserPointer.cs:line 120
   at WlxOverlay.Overlays.LaserPointer.AfterInput(Boolean batteryStateUpdated) in /mnt/1TB-SSD/home/redstonelp2/GIT/WlxOverlay/Overlays/LaserPointer.cs:line 70
   at WlxOverlay.Core.OverlayManager.Update() in /mnt/1TB-SSD/home/redstonelp2/GIT/WlxOverlay/Core/OverlayManager.cs:line 172
   at WlxOverlay.GFX.OpenGL.GlGraphicsEngine.OnRender(Double _) in /mnt/1TB-SSD/home/redstonelp2/GIT/WlxOverlay/GFX/OpenGL/GlGraphicsEngine.cs:line 68
   at Silk.NET.Windowing.Internals.ViewImplementationBase.DoRender()
   at Silk.NET.Windowing.WindowExtensions.<>c__DisplayClass2_0.<Run>b__0()
   at Silk.NET.Windowing.Internals.ViewImplementationBase.Run(Action onFrame)
   at Silk.NET.Windowing.WindowExtensions.Run(IView view)
   at WlxOverlay.GFX.OpenGL.GlGraphicsEngine.StartEventLoop() in /mnt/1TB-SSD/home/redstonelp2/GIT/WlxOverlay/GFX/OpenGL/GlGraphicsEngine.cs:line 41
   at Program.<Main>$(String[] args) in /mnt/1TB-SSD/home/redstonelp2/GIT/WlxOverlay/Program.cs:line 120
   at Program.<Main>(String[] args)
[1]    43018 IOT instruction (core dumped)  ./WlxOverlay
```

**Comment by galister** [comment]
ok, try apply the fix from f215ad5

**Comment by RedstoneLP2** [comment]
Running straight from master made me realize i didn't update for quite a while... anyway, the offset seems to be gone now
